### PR TITLE
[Self-service error codes] Update existing error code references [DPP-593]

### PR DESCRIPTION
**Pull request by tudor-da**
Updated pre-existing gRPC status code details in Daml docs.
Additionally:
* small re-wordings in the Error Codes index page.
* moved the error codes section under Building Applications -> Ledger API -> gRPC 

**Hints for review**: Check the file diffs. Preview: https://htmlpreview.github.io/?https://raw.githubusercontent.com/tudor-da/docs-daml/master/gen/app-dev/grpc/error-codes.html
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/app-dev/grpc/error-codes.rst [hunk 1]
```diff
@@ -1,27 +1,26 @@
 .. Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
 .. SPDX-License-Identifier: Apache-2.0
 
+.. _error-codes:
```
Comment (pbatko-da) [inline docs/source/app-dev/grpc/error-codes.rst:4]: What does this do?

File: docs/source/app-dev/grpc/index.rst [hunk 2]
```diff
@@ -66,19 +73,31 @@ For information on how Daml types and contracts are represented by the Ledger AP
 Error handling
 **************
 
-For the standard error codes that the server or the client might return, see the `gRPC documentation <https://github.com/grpc/grpc/blob/600272c826b48420084c2ff76dfb0d34324ec296/doc/statuscodes.md>`__ .
+The Ledger API generally uses the gRPC standard error codes for signaling response failures to client applications.
```
Comment (pbatko-da) [inline docs/source/app-dev/grpc/index.rst:76]:
```suggestion
The Ledger API generally uses the gRPC standard status codes for signaling response failures to client applications.
```

File: docs/source/app-dev/grpc/index.rst [hunk 3]
```diff
@@ -66,19 +73,31 @@ For information on how Daml types and contracts are represented by the Ledger AP
...
+
+For more details on the gRPC standard error codes, see the `gRPC documentation <https://github.com/grpc/grpc/blob/600272c826b48420084c2ff76dfb0d34324ec296/doc/statuscodes.md>`__ .
```
Comment (pbatko-da) [inline docs/source/app-dev/grpc/index.rst:78]:
```suggestion
For more details on the gRPC standard status codes, see the `gRPC documentation <https://github.com/grpc/grpc/blob/600272c826b48420084c2ff76dfb0d34324ec296/doc/statuscodes.md>`__ .
```

File: docs/source/app-dev/grpc/index.rst [hunk 4]
```diff
@@ -66,19 +73,31 @@ For information on how Daml types and contracts are represented by the Ledger AP
...
 
-For submitted commands, there are these response codes:
+Generically, on submitted commands the Ledger API responds with the following gRPC status codes:
```
Comment (pbatko-da) [inline docs/source/app-dev/grpc/index.rst:80]: Could this section be dropped and instead we should point to the error categories inventory?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto [hunk 5]
```diff
@@ -54,6 +51,7 @@ service PartyManagementService {
   // - ``UNAUTHENTICATED``: if the request does not include a valid access token
   // - ``PERMISSION_DENIED``: if the claims in the token are insufficient to perform a given operation
   // - ``UNIMPLEMENTED``: if synchronous party allocation is not supported by the backing participant
+  // - ``DEADLINE_EXCEEDED``: if the request times out
```
Comment (pbatko-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto:54]: Going forward it seems to me we want to thing of signalling errors in terms of error categories, not gRpc status codes. If so, then at some point it could be easier, from maintenance standpoint, to list error categories here instead.

**Review by pbatko-da — APPROVED** [review]

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/package_management_service.proto [hunk 6]
```diff
@@ -11,9 +11,6 @@ option java_outer_classname = "PackageManagementServiceOuterClass";
 option java_package = "com.daml.ledger.api.v1.admin";
 option csharp_namespace = "Com.Daml.Ledger.Api.V1.Admin";
 
-// Status: experimental interface, will change before it is deemed production
```
Comment (mziolekda) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/package_management_service.proto:14]: Please keep this comment.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto [hunk 7]
```diff
@@ -10,9 +10,6 @@ option java_outer_classname = "PartyManagementServiceOuterClass";
 option java_package = "com.daml.ledger.api.v1.admin";
 option csharp_namespace = "Com.Daml.Ledger.Api.V1.Admin";
 
-// Status: experimental interface, will change before it is deemed production
```
Comment (mziolekda) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto:13]: Please keep this comment

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto [hunk 8]
```diff
@@ -9,9 +9,6 @@ option java_outer_classname = "ParticipantPruningServiceOuterClass";
 option java_package = "com.daml.ledger.api.v1.admin";
 option csharp_namespace = "Com.Daml.Ledger.Api.V1.Admin";
 
-// Status: experimental interface, will change before it is deemed production
```
Comment (mziolekda) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto:12]: I think is fine to remove this, however, I would not want to do it as part of this PR. I would like his to be discussed and removed on a separate PR.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_completion_service.proto [hunk 9]
```diff
@@ -33,9 +33,10 @@ service CommandCompletionService {
   // Errors:
   // - ``UNAUTHENTICATED``: if the request does not include a valid access token
   // - ``PERMISSION_DENIED``: if the claims in the token are insufficient to perform a given operation
-  // - ``NOT_FOUND``: if the request does not include a valid ledger id or if the ledger has been pruned before ``begin``
+  // - ``NOT_FOUND``: if the request does not include a valid ledger id
+  // - ``FAILED_PRECONDITION``: if the ledger has been pruned after the subscription start offset
   // - ``INVALID_ARGUMENT``: if the payload is malformed or is missing required fields
-  // - ``OUT_OF_RANGE``: if the absolute offset is not before the end of the ledger
+  // - ``OUT_OF_RANGE``: if the absolute offset is after the end of the ledger
```
Comment (mziolekda) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_completion_service.proto:39]: Have we changed the semantics of this error? Please note that `not before` is not the same as `after`.

Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_completion_service.proto:39]: The semantics haven't changed in this PR. This a correction to reflect the status quo actually. See https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala#L55-L56 which eventually dispatches to https://github.com/digital-asset/daml/blob/main/ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/LedgerOffsetValidator.scala#L59-L63

Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_completion_service.proto:39]: This mismatch was there for a while afaict, since the logic using `gt` is there from at least mid 2020

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_service.proto [hunk 10]
```diff
@@ -32,8 +32,8 @@ service TransactionService {
   // Errors:
   // - ``UNAUTHENTICATED``: if the request does not include a valid access token
   // - ``PERMISSION_DENIED``: if the claims in the token are insufficient to perform a given operation
-  // - ``NOT_FOUND``: if the request does not include a valid ledger id or if the ledger has been pruned before ``begin``
-  // - ``INVALID_ARGUMENT``: if the payload is malformed or is missing required fields (e.g. if ``before`` is not before ``end``)
```
Comment (mziolekda) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_service.proto:36]: This error seems to still apply see e.g. missingField reachable from commonValidations

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_completion_service.proto (hunk 9 above)
Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_completion_service.proto:39]: :+1:

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_service.proto (hunk 10 above)
Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_service.proto:36]: True. Reverting

File: docs/source/app-dev/grpc/error-codes.rst (hunk 1 above)
Comment (tudor-da) [reply docs/source/app-dev/grpc/error-codes.rst:4]: So this page can be referenced from other places

Comment (tudor-da) [reply docs/source/app-dev/grpc/error-codes.rst:4]: But since the file name is the same might not be needed. Let me check

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto (hunk 7 above)
Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto]: Reverted

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto (hunk 8 above)
Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto]: Reverted

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/package_management_service.proto (hunk 6 above)
Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/package_management_service.proto]: Reverted

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_service.proto (hunk 10 above)
Comment (tudor-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_service.proto:36]: Done

**Review by mziolekda — APPROVED** [review]
Looks ok now!
